overcooked_ai_py/mdp/overcooked_mdp.py

```diff
@@ -1600,24 +1614,43 @@ def process_for_player(primary_agent_idx):
             # OBJECT & STATE LAYERS
             for obj in all_objects:
                 if obj.name == "soup":
-                    if Recipe.ONION in obj.ingredients:
-                        if obj.position in self.get_pot_locations():
-                            state_mask_dict["onions_in_pot"] += make_layer(obj.position, len(obj.ingredients))
-                            state_mask_dict["onions_cook_time"] += make_layer(obj.position, max(0, obj._cooking_tick))
+                    # removed the next line because onion doesn't have to be in all the soups?
+                    # if Recipe.ONION in obj.ingredients:
+                    # get the ingredients into a {object: number} dictionary
+                    ingredients_dict = Counter(obj.ingredients)
+                    # assert "onion" in ingredients_dict.keys()
+                    if obj.position in self.get_pot_locations():
+                        if obj.is_idle:
+                            # onions_in_pot and tomatoes_in_pot are used when the soup is idling, and ingredients could still be added
+                            state_mask_dict["onions_in_pot"] += make_layer(obj.position, ingredients_dict["onion"])
+                            state_mask_dict["tomatoes_in_pot"] += make_layer(obj.position, ingredients_dict["tomato"])
                         else:
-                            # If player soup is not in a pot, put it in separate mask
-                            state_mask_dict["onion_soup_loc"] += make_layer(obj.position, 1)
+                            state_mask_dict["onions_in_soup"] += make_layer(obj.position, ingredients_dict["onion"])
+                            state_mask_dict["tomatoes_in_soup"] += make_layer(obj.position, ingredients_dict["tomato"])
+                            state_mask_dict["soup_cook_time_remaining"] += make_layer(obj.position, obj.cook_time_remaining)
```

Comment:
There is no `cook_time_remaining` property in the SoupState class, but it would be super easy to add if it would make it syntatically easier for you. I've just been using `cook_time_remaining = soup.cook_time - soup._cooking_tick`

Reply:
I added a function since I was calling the same property somewhere else. I am adding it back in the newest commit